### PR TITLE
HDDS-9122. Make the Datanode CloseContainerCommandHandler async by queuing commands in an executor

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

From logs, its observed that command is getting queued up. And as per code, all handlers are asynchronous except closeContainer. And in code, found 2 of possibility:
1. container write lock stuck while mark container close
2. ratis submit -- may be possible to block -- not found much info

Additionally, when DN issue, there will be quite a lot of close container commands to be executed, may be 100s of open container also blocking further request.

All the Handlers are executed async way, so same is added for close container. 
As solution,
- close container command is executed async way with 3 thread in parallel with queue size of 5000 (default).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9122

## How was this patch tested?

Existing integration testcase cover the close container behavior

### CODE REVIEW

**Comment by sodonnel** [comment]
Please add an outline of the problem, the symptoms, and how the solution addresses it to the PR description.

I think I had looked briefly at this problem before - it was not clear to me that the close container commands were so slow it was affecting things, especially since the problem was detected on a cluster with a very small number of containers. Were you able to prove the slow container closes were the root cause of the problem via debug logging, jstacks etc?

It does make sense to make this process async and not be able to block the command handler thread, but I am interested to know if this is the root cause of the problem or is it possible that some other command is blocking the queue preventing the container close commands from getting executed?

**Comment by sumitagrawl** [comment]
> Please add an outline of the problem, the symptoms, and how the solution addresses it to the PR description.
> 
> I think I had looked briefly at this problem before - it was not clear to me that the close container commands were so slow it was affecting things, especially since the problem was detected on a cluster with a very small number of containers. Were you able to prove the slow container closes were the root cause of the problem via debug logging, jstacks etc?
> 
> It does make sense to make this process async and not be able to block the command handler thread, but I am interested to know if this is the root cause of the problem or is it possible that some other command is blocking the queue preventing the container close commands from getting executed?

Updated...

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 1]
```diff
@@ -130,6 +133,19 @@ public class DatanodeConfiguration {
...
 
+  /**
+   * The maximum number of threads used to close containers on a datanode
+   * simultaneously.
+   */
+  @Config(key = "container.close.threads.max",
+      type = ConfigType.INT,
+      defaultValue = "3",
+      tags = {DATANODE},
+      description = "The maximum number of threads used to delete containers " +
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:144]: Typo - Delete -> Close

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CloseContainerCommandHandler.java [hunk 2]
```diff
@@ -69,73 +85,76 @@ public CloseContainerCommandHandler() {
...
-        if (closeCommand.getForce()) {
-          controller.closeContainer(containerId);
+
+        // move the container to CLOSING if in OPEN state
+        controller.markContainerForClose(containerId);
+
+        switch (container.getContainerState()) {
+        case OPEN:
+        case CLOSING:
+          // If container is part of open pipeline, close it via write channel
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CloseContainerCommandHandler.java:115]: Nit: Removing "the" from the sentence, makes it incorrect. The sentence should read:

> If the container is part of **an** open pipeline, close it via write channel

The typo is the missing "an" if you want to fix it but removing "the" is not correct

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CloseContainerCommandHandler.java [hunk 3]
```diff
@@ -69,73 +85,76 @@ public CloseContainerCommandHandler() {
...
+          if (ozoneContainer.getWriteChannel()
+              .isExist(closeCommand.getPipelineID())) {
+            ContainerCommandRequestProto request =
+                getContainerCommandRequestProto(datanodeDetails,
+                    closeCommand.getContainerID(),
+                    command.getEncodedToken());
+            ozoneContainer.getWriteChannel()
+                .submitRequest(request, closeCommand.getPipelineID());
+          } else if (closeCommand.getForce()) {
+            // Non-RATIS containers should have force close flag set, so they
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CloseContainerCommandHandler.java:125]: Nit: Same as above - we should not remove "the" and "than"

**Review by sodonnel — APPROVED** [review]

**Pull request merged**